xfrd-tcp.c

```diff
@@ -989,9 +1309,19 @@ xfrd_tcp_pipe_release(struct xfrd_tcp_set* set, struct xfrd_tcp_pipeline* tp,
 		event_del(&tp->handler);
 	tp->handler_added = 0;
 
+#ifdef HACVE_TLS_1_3
+	/* close SSL */
+	if (tp->ssl) {
+		DEBUG(DEBUG_XFRD, 1, (LOG_INFO, "xfrd: Shutting down TLS"));
+		SSL_shutdown(tp->ssl);
+		SSL_free(tp->ssl);
+		tp->ssl = NULL;
+	}
+#endif
+
 	/* fd in tcp_r and tcp_w is the same, close once */
 	if(tp->tcp_r->fd != -1)
-		close(tp->tcp_r->fd);
+		shutdown(tp->tcp_r->fd, SHUT_RDWR);
```

Comment:
This is the wrong code? It has to close the fd, otherwise it is leaking the file descriptor.  Regardless if you want to call shutdown before close, the close itself has to happen.  Why do you want to call the shutdown? I see no reason for it, and would put the close back again.  Perhaps it was because of the wrong ifdef above on the SSL_shutdown, that a shutdown here gave a nicer result on the other end of the connection, what with not closed but shutdown.

Reply:
Hi Wouter, we initially had close instead of shutdown, but every close caused TCP RSTs. Agreed that we should close the fd... but am trying to figure out how to prevent the RSTs. Perhaps we should do a read before the close?

I've incorporated all the rest of your suggestions except that I'll wait for Sara to weigh in on the open TODO since she found it - will also try to recreate it tomorrow.

Reply:
Yes I would also like to see it work well in practice.  But it sounds like the SSL_shutdown may do the protocol TLS shutdown negotiation, and thus prevent future packets, and thus prevent the TCP RSTs in response to those packets.  And that SSL_shutdown was commented out by the ifdef spelling error in the code above this.  So perhaps with that fix the RSTs are gone?  A read before close does not sound like a solution I have seen elsewhere.

The open issue sounds like something that is a bug, good you can recreate it, that helps figuring out what is going on with the various return states at the time when it goes wrong.

Reply:
Even with the directive correctly formed I get the TCP RSTs.

Reply:
I [asked around](https://stackoverflow.com/questions/66484149/how-do-i-prevent-tcp-resets-after-closing-file-descriptor-that-ive-already-call) and read Bert Hubert's [article](https://blog.netherlabs.nl/articles/2009/01/18/the-ultimate-so_linger-page-or-why-is-my-tcp-not-reliable) on how to handle TCP close gracefully, but even after adding a `read()` and check for 0 before `close(fd)`, I get RSTs... not sure how to proceed.

Reply:
Since NSD is the client here for receiving AXFR data.  If it wants to close, it can close.  Even if there is pending data.  TCP or TLS.  Data that does not arrive is not NSD's problem, because it wants to close.  So it is fine, the other side receiving RST stops sending the AXFR contents, also no problem.  So I do not think the RST is causing an operational problem here.  Just call close().

Reply:
I see this is `close()` now in current code, so resolved.